src/TokenBuyer.sol

```diff
@@ -0,0 +1,319 @@
+     */
+    function buyETH(uint256 tokenAmount) external nonReentrant whenNotPaused {
+        uint256 amount = Math.min(tokenAmount, tokenAmountNeeded());
+
+        paymentToken.safeTransferFrom(msg.sender, payer, amount);
+
+        safeSendETH(msg.sender, ethAmountPerTokenAmount(amount), '');
+    }
+
+    function buyETH(
+        uint256 tokenAmount,
+        address to,
+        bytes calldata data
+    ) external nonReentrant whenNotPaused {
+        uint256 amount = Math.min(tokenAmount, tokenAmountNeeded());
+        uint256 balanceBefore = paymentToken.balanceOf(address(this));
+
+        safeSendETH(to, ethAmountPerTokenAmount(amount), abi.encode(msg.sender, amount, data));
+
+        uint256 tokensReceived = paymentToken.balanceOf(address(this)) - balanceBefore;
+        if (tokensReceived < amount) {
+            revert ReceivedInsufficientTokens(amount, tokensReceived);
+        }
+
+        paymentToken.safeTransfer(payer, tokensReceived);
+    }
+
+    /**
+     * @notice Allow ETH top-ups.
+     */
+    receive() external payable {}
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      VIEW FUNCTIONS
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /**
+     * @notice Get how much ETH this contract needs in order to fund its current obligations plus `additionalTokens`, with
+     * a safety buffer `bufferBPs` basis points.
+     * @param additionalTokens an additional amount of `paymentToken` liability to use in this ETH requirement calculation, in payment token decimals.
+     * @param bufferBPs the number of basis points to add on top of the token liability price in ETH as a safety buffer, e.g.
+     * if `bufferBPs` is 10K, the function will return twice the amount it needs according to price alone.
+     * @return uint256 the amount of ETH needed to fund `additionalTokens` with a `bufferBPs` safety buffer.
+     */
+    function ethNeeded(uint256 additionalTokens, uint256 bufferBPs) public view returns (uint256) {
+        uint256 tokenAmount = tokenAmountNeeded() + additionalTokens;
+        uint256 ethCostOfTokens = ethAmountPerTokenAmount(tokenAmount);
+        uint256 ethCostWithBuffer = (ethCostOfTokens * (bufferBPs + 10_000)) / 10_000;
+
+        return ethCostWithBuffer - address(this).balance;
+    }
+
+    function tokenAmountNeededAndETHPayout() public view returns (uint256 tokenAmount, uint256 ethAmount) {
+        tokenAmount = tokenAmountNeeded();
+        ethAmount = ethAmountPerTokenAmount(tokenAmount);
+    }
+
+    /**
+     * @return uint256 the amount of `paymentToken` this contract is willing to buy in exchange for ETH, in payment token decimals.
+     */
+    function tokenAmountNeeded() public view returns (uint256) {
+        uint256 _paymentTokenBalance = paymentToken.balanceOf(address(this));
+        uint256 iouSupply = iouToken.totalSupply();
+        unchecked {
+            if (_paymentTokenBalance > baselinePaymentTokenAmount + iouSupply) {
```

Comment:
maybe can be shortened to:
`return Math.max(baselinePaymentTokenAmount + iouSupply - _paymentTokenBalance, 0)`

don't think it will have higher gas cost, maybe less?

Reply:
I believe if `_paymentTokenBalance` is greater than `baselinePaymentTokenAmount + iouSupply` the result of the calculation you're suggesting would be an overflow, no?

Reply:
hmm yeah it will underflow, so maybe need to be cast to signed int..
or we can just ignore this

Reply:
resolving this thread :)